rclcpp/include/rclcpp/node_handle.hpp

```diff
@@ -0,0 +1,280 @@
+  inline BaseInterface base() {return base_;}
+
+  /// Set the bound NodeBaseInterface
+  inline void base(BaseInterface interface) {base_ = interface;}
+
+  /// Return the bound NodeClockInterface
+  inline ClockInterface clock() {return clock_;}
+
+  /// Set the bound NodeClockInterface
+  inline void clock(ClockInterface interface) {clock_ = interface;}
+
+  /// Return the bound NodeGraphInterface
+  inline GraphInterface graph() {return graph_;}
+
+  /// Set the bound NodeGraphInterface
+  inline void graph(GraphInterface interface) {graph_ = interface;}
+
+  /// Return the bound NodeLoggingInterface
+  inline LoggingInterface logging() {return logging_;}
+
+  /// Set the bound NodeLoggingInterface
+  inline void logging(LoggingInterface interface) {logging_ = interface;}
+
+  /// Return the bound NodeTimersInterface
+  inline TimersInterface timers() {return timers_;}
+
+  /// Set the bound NodeTimersInterface
+  inline void timers(TimersInterface interface) {timers_ = interface;}
+
+  /// Return the bound NodeTopicsInterface
+  inline TopicsInterface topics() {return topics_;}
+
+  /// Set the bound NodeTopicsInterface
+  inline void topics(TopicsInterface interface) {topics_ = interface;}
+
+  /// Return the bound NodeServicesInterface
+  inline ServicesInterface services() {return services_;}
+
+  /// Set the bound NodeServicesInterface
+  inline void services(ServicesInterface interface) {services_ = interface;}
+
+  /// Return the bound NodeWaitablesInterface
+  inline WaitablesInterface waitables() {return waitables_;}
+
+  /// Set the bound NodeWaitablesInterface
+  inline void waitables(WaitablesInterface interface) {waitables_ = interface;}
+
+  /// Return the bound NodeParametersInterface
+  inline ParametersInterface parameters() {return parameters_;}
+
+  /// Set the bound NodeParametersInterface
+  inline void parameters(ParametersInterface interface) {parameters_ = interface;}
+
+  /// Return the bound NodeTimeSourceInterface
+  inline TimeSourceInterface time_source() {return time_source_;}
+
+  /// Set the bound NodeTimeSourceInterface
+  inline void time_source(TimeSourceInterface interface) {time_source_ = interface;}
+
+
+  // Alterate getters
+  inline BaseInterface get_node_base_interface() {return base();}
+  inline ClockInterface get_node_clock_interface() {return clock();}
+  inline GraphInterface get_node_graph_interface() {return graph();}
+  inline LoggingInterface get_node_logging_interface() {return logging();}
+  inline TimersInterface get_node_timers_interface() {return timers();}
+  inline TopicsInterface get_node_topics_interface() {return topics();}
+  inline ServicesInterface get_node_services_interface() {return services();}
+  inline WaitablesInterface get_node_waitables_interface() {return waitables();}
+  inline ParametersInterface get_node_parameters_interface() {return parameters();}
+  inline TimeSourceInterface get_node_time_source_interface() {return time_source();}
```

Comment:
Why do we need alternate getters?

Also why don't we have something like this instead:

```c++
node_handle.get<NodeBaseInterface>();
```

Reply:
One set is for convenience and the other set is to mimic the node interface getters. If I had to choose between the three (the two I mentioned earlier, and the `get<>()`), I think I'd go with the node interface getters, for parity with node-like types :o

Reply:
The current status of this is that the getters have all been removed on the aggregating `NodeInterfaces` class and are instead implemented by each of the support classes that the `NodeInterfaces` object inherits from (via multiple inheritance).

E.g. https://github.com/ros2/rclcpp/pull/2041/files#diff-10a9fcfcaf4778e4e836e88ad35ed243e3087b815bb2b8f46fc786042b3533f4R79

While it's true that the support classes create a bit of boilerplate, they're what's necessary to allow extensions to new types in the future (as well as supporting NodeTs with unconventional getter names for the interfaces (in which case, a new support class can be written by the user for that NodeT.))

Specifically, it allows users to specify:
- [How to get the interface from the passed in NodeT](https://github.com/ros2/rclcpp/pull/2041/files#diff-10a9fcfcaf4778e4e836e88ad35ed243e3087b815bb2b8f46fc786042b3533f4R53)
- [How users are expected to get the interface out from the NodeInterface](https://github.com/ros2/rclcpp/pull/2041/files#diff-10a9fcfcaf4778e4e836e88ad35ed243e3087b815bb2b8f46fc786042b3533f4R56)
- [And how to rebind the interface](https://github.com/ros2/rclcpp/pull/2041/files#diff-10a9fcfcaf4778e4e836e88ad35ed243e3087b815bb2b8f46fc786042b3533f4R59) (though this is probably getting removed) 

---

With the current status then, I'm not sure if a `get<>()` would be able to remove boilerplate compared to the current implementation or we need to constrain the support classes, which reduces extensibility.

In the former case we'll need a get<>() specialization on each support class to define how to access the interface that they store (which is really just renaming the current getter names on each support class..), and in the latter case we need to enforce that every support class has a consistent API (either a static getter name, a static protected/public member) so that we can remove the getters on each support class and have a single overarching one on the NodeInterface.

I think between the two options, I rather just leave it up to the user to define how they want to do their getters (but use the current getter naming convention for the current node interfaces.)

Reply:
Another thing I was considering was using a tuple to store the interfaces, but that doesn't work because:
- There still needs to be a way to define how to obtain interfaces from a passed in NodeT in a variadic manner
- It means we can't aggregate more than one instance of an identical interface type (e.g. two NodeBaseInterfaces) (with support classes you could do it by defining two different support classes that both store NodeBaseInterfaces)

The only way to make tuples work is to remove the NodeT constructor, which I am very hesitant to do because it significantly cuts down on the amount of code a user has to write to pass around nodes/lifecycle nodes to a method.